pgjdbc/src/test/java/org/postgresql/test/jdbc2/BinaryTest.java

```diff
@@ -0,0 +1,282 @@
+   */
+  @After
+  public void tearDown() throws SQLException {
+    TestUtil.closeDB(con);
+
+    con = TestUtil.openDB();
+
+    TestUtil.dropTable(con, "test_binary");
+
+    TestUtil.closeDB(con);
+  }
+
+  /**
+   * Make sure the functions for adding binary transfer OIDs for custom types are correct.
+   *
+   * @throws SQLException if a database error occurs
+   */
+  @Test
+  public void testBinaryTransferOids() throws SQLException {
+    con = TestUtil.openDB();
+    QueryExecutor queryExecutor = ((PgConnection) con).getQueryExecutor();
+    // get current OIDs (make a copy of them)
+    Set<Integer> oidsReceive = new HashSet<Integer>(queryExecutor.getBinaryReceiveOids());
+    Set<Integer> oidsSend = new HashSet<Integer>(queryExecutor.getBinarySendOids());
+    // add a new OID to be transferred as binary data
+    int customTypeOid = 91716;
+    // first for receiving
+    queryExecutor.addBinaryReceiveOid(91716);
+    // check new OID
+    assertTrue(queryExecutor.useBinaryForReceive(customTypeOid));
+    assertFalse(queryExecutor.useBinaryForSend(customTypeOid));
+    // make sure all other OIDs are still there
+    for (int oid : oidsReceive) {
+      assertTrue(queryExecutor.useBinaryForReceive(oid));
+    }
+    for (int oid : oidsSend) {
+      assertTrue(queryExecutor.useBinaryForSend(oid));
+    }
+    // then for sending
+    queryExecutor.addBinarySendOid(91716);
+    // check new OID
+    assertTrue(queryExecutor.useBinaryForReceive(customTypeOid));
+    assertTrue(queryExecutor.useBinaryForSend(customTypeOid));
+    // make sure all other OIDs are still there
+    for (int oid : oidsReceive) {
+      assertTrue(queryExecutor.useBinaryForReceive(oid));
+    }
+    for (int oid : oidsSend) {
+      assertTrue(queryExecutor.useBinaryForSend(oid));
+    }
+    TestUtil.closeDB(con);
+  }
+
+  /**
+   * Make sure custom binary types are handled automatically.
+   *
+   * @throws SQLException if a database error occurs
+   */
+  @Test
+  public void testCustomBinaryTypes() throws SQLException {
+    con = TestUtil.openDB();
+
+    PgConnection pgconn = (PgConnection) con;
+    QueryExecutor queryExecutor = pgconn.getQueryExecutor();
+
+    // define an oid of a binary type for testing, POINT is used here as it already exists in the
+    // database and requires no complex own type definition
+    int customTypeOid = Oid.POINT;
+    // the only disadvantage is that it is added automatically in the PgConnection, so be sure to
+    // remove type OID from predefined lists
+    queryExecutor.getBinaryReceiveOids().remove(customTypeOid);
+    queryExecutor.getBinarySendOids().remove(customTypeOid);
+    assertFalse(queryExecutor.useBinaryForReceive(customTypeOid));
+    assertFalse(queryExecutor.useBinaryForSend(customTypeOid));
+
+    // now define a custom type
+    pgconn.addDataType("point", TestCustomType.class);
+    // check if it was automatically added
+    assertTrue(queryExecutor.useBinaryForReceive(customTypeOid));
```

Comment:
Can you please clarify what automatically adds the binary mode for POINT? It is hard to understand here.

Reply:
I added a message to the assert. Basically the new functionality adds types implementing `PGBinaryObject` to the types to be binary received and sent.